### PR TITLE
Enhance Datepicker be extensible by users to add custom buttons with user specified dates

### PR DESCRIPTION
**Pull request by hsiddanthi**
Enhances Datepicker to allow user to extend picker by adding custom buttons with specified dates, like for example Yesterday, Start/End of Quarter, Birthdays :) and many more, user provided dates. Allows users to add more than one button, though not recommanded to have more than 2 for better UX.

This PR includes the changes to

1. Datepicker and DateRangepicker API to allow users to add button with label and function that provides the date.
2. Add implementation in Picker UI and a place holder for these buttons in pickerTemplate and add new quick control template
3. Add test cases to Datepicker to test new API
4. Styling to these new buttons, similar styling as Clear/Today button to be consistant with rest of UI
5. Finally, add them to demo code that demonstrates this new feature

Potential issues not tested are buttons are diplayed at top when Picker pops up than down; and when user provided function takes long time to return a date (calls a remote server etc.), could degrade the performance of Picker UI. Considering the simple scenarios of user function returns quickly returns outweights these issues, hopefully.

### CODE REVIEW

**Comment by hsiddanthi** [comment]
Hell, I am new to Github PRs, so it added both individual commits (one bug fix and another enhancement), anyways here is the details on enhancement:

**Enhance Datepicker be extensible by users to add custom buttons with user specified dates**

Enhances Datepicker to allow user to extend picker by adding custom buttons with specified dates, like for example Yesterday, Start/End of Quarter, Birthdays :) and many more, user provided dates. Allows users to add more than one button, though not recommanded to have more than 2 for better UX.

This PR includes the changes to

1. Datepicker and DateRangepicker API to allow users to add button with label and function that provides the date.
2. Add implementation in Picker UI and a place holder for these buttons in pickerTemplate and add new quick control template
3. Add test cases to Datepicker to test new API
4. Styling to these new buttons, similar styling as Clear/Today button to be consistant with rest of UI
5. Finally, add them to demo code that demonstrates this new feature

Potential issues not tested are buttons are diplayed at top when Picker pops up than down; and when user provided function takes long time to return a date (calls a remote server etc.), could degrade the performance of Picker UI. Considering the simple scenarios of user function returns quickly returns outweights these issues, hopefully.

**Comment by mymth** [comment]
You need to think about the cases where 1) null or an empty string is passed to btnLabel and 2) addQuickControl() is called multiple times with the same btnLabel before calling removeQuickControl().
For No. 1, if you want to prevent buttons from becoming label-less, you need to check null and empty string too because they also make the button’s label empty. If not, then you don’t need to check if btnLabel is undefined.
For No. 2, the addQuickControl() method allows users to have multiple buttons with the same label and the removeQuickControl() method removes all of them in one call. Though it won’t cause serious problem in real-world use cases, it’s better to avoid because, in this pattern of programs, most programmers would expect the label to be a key.

Also, the name of addQuickControl()’s second argument is problematic because it’s the same as the other function used for actual event listener. The function expected for it is not the click listener of the button but one to provide the date to set on button click. The name should represent its functionality for better readability.

But more fundamentally, I don’t like where and how buttons are added. Having additional buttons on topside apart from the today/clear buttons doesn't look a good UI to me. In fact, while trying out the feature, I mistakenly clicked a quick control button a few times attempting to change the view to month/year. And when I added multiple buttons, full-width buttons were piled up on the top of the picker. It looks to me that, due to the increase in the distance between the calendar area and the input field, having 2 or more buttons sacrifices the usability of date picker’s basic features and isn't really functional.

The same enhancement as this PR can be done by creating a subclass like this: https://codepen.io/mymth/pen/OJOMPqN. If you don’t need to replace/remove buttons (I think it's more common), adding buttons ad hoc like this: https://codepen.io/mymth/pen/XWzNpVx should work too.
I have no intention of making this library feature rich. I'd rather keep my focus on the quality of core functioanlity. If some new feature that improves core functionality requires changes in the library’s code, I’ll be happy to add it. But as for convenient features that can be made without touching the library’s code (like this), I’d like to leave them in users’ space.

In regard to showing the picker when the input field has readonly attribute, it’s actually intentional and an inheritance from bootstrap-datepicker’s default behavior. I actually thought about changing it like you change but decided not to do so because it seems to  be better to provide a way to change the date picker’s behavior to allow users only to select date(s) from the calendar, like flatpickr.
However I see 2 problems. One is that the picker enters edit mode even though input field is not editable. The other is that users have no choice—omitting enableOnReadonly from the options ported from bootstrap-datepicker was a mistake.
So I’ll fix these in the next version.

In conclusion, I'm very sorry but for the above reasons, I will turn down this PR.
Thank you for your understanding.

**Comment by hsiddanthi** [comment]
Thanks for review and feedback, I will go with extending the library. I totally understand and respect that you would like to keep the library simple and that's the reason I chose this in my project.

Though, I would like to understand more on when setting the date with `setDate` funtion, would expect to auto-refresh than seperate call, may be that can be default in `setDate` function?

`// refresh the picker separately to ensure it goes back to the selected date
      this.setDate(getDateToSet(), {render: false});
      this.refresh('picker');`
